### PR TITLE
Refactoring: more generic intervals support and proper semi-closed intervals usage

### LINKED ISSUE #268: Support for off-chain validation and intervals for `AuctionStage` and fix boundary case
Couple of Tx do not fall into single stage. This requires hardcoding of intervals. Support for intervals over `AuctionStage` would cover this.

Also currently intervals for `Tx.*` are created manually.

- [x] Fix boundary case (@uhbif19 has draft for that locally)
- [ ] Support intervals of stages (or maybe just stage-to-infinity case, which is only used in our code)
- [x] Support Tx `validityBound` construction
- [x] Refactor current code to use this

### PR DESCRIPTION
**Pull request by nazrhom**

### CODE REVIEW

File: src/HydraAuction/OnChain/Common.hs [hunk 1]
```diff
@@ -30,12 +30,20 @@ minAuctionFee = 2_000_000
...
-  BiddingEndedStage -> interval (biddingEnd terms) (voucherExpiry terms)
-  VoucherExpiredStage -> interval (voucherExpiry terms) (cleanup terms)
+  AnnouncedStage -> strictTo (biddingStart terms)
+  BiddingStartedStage -> rightExclusiveInterval (biddingStart terms) (biddingEnd terms)
+  BiddingEndedStage -> rightExclusiveInterval (biddingEnd terms) (voucherExpiry terms)
+  VoucherExpiredStage -> rightExclusiveInterval (voucherExpiry terms) (cleanup terms)
   CleanupStage -> from (cleanup terms)
 
+{-# INLINEABLE strictTo #-}
+strictTo :: a -> Interval a
```
Comment (uhbif19) [inline src/HydraAuction/OnChain/Common.hs]: Move this stuff to utils, please.

File: src/HydraAuction/Tx/Deposit.hs [hunk 2]
```diff
@@ -161,7 +163,7 @@ losingBidderClaimDeposit terms = do
         , outs = []
         , toMint = TxMintValueNone
         , changeAddress = bidderAddress
-        , validityBound = (Just $ biddingEnd terms, Nothing)
+        , validityBound = stageToInterval terms BiddingEndedStage
```
Comment (uhbif19) [inline src/HydraAuction/Tx/Deposit.hs]: This is incorrect.

Left logic has no upper bound, while stage has.

File: src/HydraAuction/Tx/Deposit.hs [hunk 3]
```diff
@@ -192,7 +194,7 @@ sellerClaimDepositFor terms bidderPkh = do
         , outs = []
         , toMint = TxMintValueNone
         , changeAddress = sellerAddress
-        , validityBound = (Just $ voucherExpiry terms, Nothing)
+        , validityBound = stageToInterval terms VoucherExpiredStage
```
Comment (uhbif19) [inline src/HydraAuction/Tx/Deposit.hs]: Same.

File: src/HydraAuction/Tx/Escrow.hs [hunk 4]
```diff
@@ -364,5 +366,5 @@ sellerReclaims terms = do
             ]
         , toMint = TxMintValueNone
         , changeAddress = sellerAddress
-        , validityBound = (Just $ voucherExpiry terms, Nothing)
+        , validityBound = stageToInterval terms VoucherExpiredStage
```
Comment (uhbif19) [inline src/HydraAuction/Tx/Escrow.hs]: Same.

File: src/HydraAuction/Tx/Deposit.hs (hunk 2 above)
Comment (uhbif19) [reply src/HydraAuction/Tx/Deposit.hs]: Solution I had in mind was intervals over stage, maybe I formulate that in issue description badly.

While seems like stage-to-posinf is the only case we do use.

File: src-utils/HydraAuctionUtils/Tx/AutoCreateTx.hs [hunk 5]
```diff
@@ -91,7 +91,7 @@ data AutoCreateParams = AutoCreateParams
   , outs :: [TxOut CtxTx]
   , toMint :: TxMintValue BuildTx
   , changeAddress :: Address ShelleyAddr
-  , validityBound :: (Maybe POSIXTime, Maybe POSIXTime)
+  , validityBound :: Interval POSIXTime
```
Comment (uhbif19) [inline src-utils/HydraAuctionUtils/Tx/AutoCreateTx.hs:94]: Interesting solution.

On one side we reuse polymorphic code, on the other side Interval reuse intuitively feels problems prone to me.

**Review by uhbif19 — CHANGES_REQUESTED** [review]

File: src-utils/HydraAuctionUtils/Tx/AutoCreateTx.hs (hunk 5 above)
Comment (uhbif19) [reply src-utils/HydraAuctionUtils/Tx/AutoCreateTx.hs:94]: Like quote from PR comment:

> and for v3 we scrap this Interval coming up with a better (non-backwards compatible) Interval definition.

Overall their backwards comparability policy does not inspire confidence, so I would not depend on Plutus more then necessary. But I guess it that case it should be fine.

**Review by uhbif19 — CHANGES_REQUESTED** [review]

File: src/HydraAuction/Tx/Deposit.hs (hunk 2 above)
Comment (nazrhom) [reply src/HydraAuction/Tx/Deposit.hs]: You are right. Maybe we can just add stage-to-posinf in utils as well, and use that?

Comment (uhbif19) [reply src/HydraAuction/Tx/Deposit.hs]: > we can just add stage-to-posinf in utils as well

But that is domain (stage) specific thing, no?

File: src/HydraAuction/OnChain/Common.hs (hunk 1 above)
Comment (nazrhom) [reply src/HydraAuction/OnChain/Common.hs:41]: Where, we don't really have onchain module in utils. I thought common onchain code should live in `onchain/common`

File: src-utils/HydraAuctionUtils/Interval.hs [hunk 6]
```diff
@@ -0,0 +1,25 @@
+{-# OPTIONS_GHC -fno-specialise #-}
```
Comment (uhbif19) [inline src-utils/HydraAuctionUtils/Interval.hs]: Sorry for nitpicking, but we have "Plutus" module in utils, so this should be its submodule as well.

File: src-utils/HydraAuctionUtils/Interval.hs [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+{-# INLINEABLE strictTo #-}
+strictTo :: a -> Interval a
+strictTo s = Interval (LowerBound NegInf True) (strictUpperBound s)
+
+{-# INLINEABLE rightExclusiveInterval #-}
+rightExclusiveInterval :: a -> a -> Interval a
+rightExclusiveInterval s s' = Interval (lowerBound s) (strictUpperBound s')
+
+{-# INLINEABLE extendIntervalRight #-}
+extendIntervalRight :: Interval a -> Interval a
```
Comment (uhbif19) [inline src-utils/HydraAuctionUtils/Interval.hs]: Not very understanible name, while I am not sure how to name it better.

**Review by uhbif19 — APPROVED** [review]

**Pull request merged**